### ISSUE TITLE
AccessHandler.editCollection: diff the old members array and the new one, and then update internalmeta on added/removed members

### ISSUE THREAD

**Issue by SaschaAdler** [crkn-rcdr/Access-Platform#84]
Brought up in #81, and I'd like to split it out into its own issue.

@RussellMcOrmond wrote:

> What `loadseries` does is a bit unusual, but might be informative.   Essentially for a given collection noid there is a desire to `add` an array of other nodes, `subtract` an array of other noides, or make the members `equal` to an array of other noids.
> 
> This isn't only managing the `members` array, but also ensuring that the changes (additions or subtractions) initiate updateInternalmeta for the members.  This will need to note which members changed and only manipulate them, and not all members of a collection (most of which won't have changed).  `loadseries` hasn't been manipulating the members of collections, but it was also intended to be a one-off tool and not regularly run as we have been doing.

Obviously the need to update a collection's members array will exist beyond the utility of `loadseries`. I haven't determined how accessor's collection API will work, exactly, but at some point it will be straightforward enough to retrieve a collection by noid and manipulate it. What does the signature of a potential `collection.updateMembers` method look like?

The metadata bus requires that any member that has been added or removed from a collection has a new `updateInternalmeta` request. For ordered collections, do members whose absolute position in the members list need to be updated? (I'm thinking of cases where we still care about `seq`).

Whatever the requirements for member updates are, it is important to avoid updating members that haven't changed; this appears to be Russell's primary concern.

I propose a `collection.updateMembers` method that receives an object argument with the following type:

```ts
interface MemberUpdateArgs {
  replace?: Noid[];
  add?: Noid[];
  remove?: Noid[];
}
```

which is to say that when calling `collection.updateMembers(args)`, args should be an object that has any of `replace`, `add`, `remove` lists of noids, and that the lists specified determine what actions are taken. If `replace` is specified, then `add` and `remove` are ignored (or maybe if either are also specified an error can be thrown). I'd like this to be one method instead of three, because it allows this action to be slightly more transactional, i.e. we won't need to update `updateInternalmeta` twice for the collection we're working on if we need to add and remove members at the same time. This method will contain the logic for updating members' `updateInternalmeta` when necessary.

I'll need to figure out the API for retrieving a collection in the first place, although it can't possibly be that complicated.

**Comment by RussellMcOrmond** [comment]
Re: " (I'm thinking of cases where we still care about seq)."

My memory is that `seq` isn't  a think in the new platform, and that sequence is entirely about the array position within the members of an ordered collection.  When looking at a manifest we would need to look to the ordered collection(s) it is a member of if we wanted to know that information.

Hammer processes a `seq` for issues, but Hammer2 does not.for manifests.


pkey and seq are pulled from the Issueinfo XML and put into the CMR-like attachment, but these aren't used for anything but display that I'm aware of, and we can drop that entirely hopefully soon.

https://github.com/crkn-rcdr/CIHM-Meta/blob/main/lib/CIHM/Meta/dmd/flatten.pm#L98-L103

**Comment by RussellMcOrmond** [comment]
Re: "Whatever the requirements for member updates are, it is important to avoid updating members that haven't changed; this appears to be Russell's primary concern."

We don't want a "small" edit done by staff to take a lot of Hammer/Press/etc processing to show up online.  So the fewer unnecessary manifest updates in relation to a collection change the better.

What you are suggesting sounds great.  Let me know how I can help, including testing/etc.

**Comment by SaschaAdler** [comment]
I'm pretty confident we don't use an issue's `seq` at all in CAP. I'm comfortable not triggering `updateInternalmeta` requests for members whose only change is their sequence in the members list.

`replace` should diff the current members list with the new one and only update members who have been added or removed.

**Comment by SaschaAdler** [comment]
Depends on #93.

**Comment by SaschaAdler** [comment]
The place to do this is in the `editCollection` method on `AccessHandler` in `couch-utils`, after the incoming data is parsed and before calling `editObject`. If the incoming data doesn't include a members array update, don't worry about it.

To start, you can get the object, parse it into a `Collection`, and retrieve its members array:

```ts
if (data.members) {
  const currentMembers = Collection.parse(await this.get(args.id)).members;
}
```

Then, compare the two arrays, and find elements that aren't in both. These are either being added or removed from the collection, and they need to have `updateInternalmeta` updated. There isn't a method yet for solely doing that, although I'll be rectifying that soon.

**Comment by SaschaAdler** [comment]
You might want to try using lodash's `xorWith` utility for this: https://lodash.com/docs#xorWith
